Handle missing build.gradle file

I have a submodule without a build.gradle file, since it has no configuration or dependencies other than what is specified in the root module build file.

This results in the following error:

> org.gradle.api.tasks.TaskValidationException: A problem was found with the configuration of task ':[module]:generateDependencyGraph'.
	at org.gradle.api.internal.tasks.execution.ValidatingTaskExecuter.report(ValidatingTaskExecuter.java:67)
	at org.gradle.api.internal.tasks.execution.ValidatingTaskExecuter.execute(ValidatingTaskExecuter.java:54)
..
	at org.gradle.internal.concurrent.ExecutorPolicy$CatchAndRecordFailures.onExecute(ExecutorPolicy.java:63)
	at org.gradle.internal.concurrent.ManagedExecutorImpl$1.run(ManagedExecutorImpl.java:46)
	at org.gradle.internal.concurrent.ThreadFactoryImpl$ManagedThreadRunnable.run(ThreadFactoryImpl.java:55)
**Caused by: org.gradle.api.InvalidUserDataException: File '/[filesystem path]/build.gradle' specified for property 'inputFile' does not exist.**
	at org.gradle.api.internal.tasks.execution.ValidatingTaskExecuter.report(ValidatingTaskExecuter.java:64)

I solved this by creating an empty build.gradle file, but keeping an empty build.gradle file smells a bit iffy.

--- a/src/main/kotlin/com/vanniktech/dependency/graph/generator/DependencyGraphGeneratorPlugin.kt
+++ b/src/main/kotlin/com/vanniktech/dependency/graph/generator/DependencyGraphGeneratorPlugin.kt
@@ -37,7 +37,6 @@ open class DependencyGraphGeneratorPlugin : Plugin<Project> {
       it.generator = this
       it.group = "reporting"
       it.description = "Generates a dependency graph$name"
-      it.inputFile = project.buildFile
       it.outputDirectory = File(project.buildDir, "reports/dependency-graph/")
     }
   }
@@ -49,7 +48,6 @@ open class DependencyGraphGeneratorPlugin : Plugin<Project> {
       it.projectGenerator = this
       it.group = "reporting"
       it.description = "Generates a project dependency graph$name"
-      it.inputFile = project.buildFile
       it.outputDirectory = File(project.buildDir, "reports/project-dependency-graph/")
     }
   }

--- a/src/main/kotlin/com/vanniktech/dependency/graph/generator/DependencyGraphGeneratorTask.kt
+++ b/src/main/kotlin/com/vanniktech/dependency/graph/generator/DependencyGraphGeneratorTask.kt
@@ -4,14 +4,12 @@ import com.vanniktech.dependency.graph.generator.DependencyGraphGeneratorExtensi
 import guru.nidi.graphviz.engine.Graphviz
 import org.gradle.api.DefaultTask
 import org.gradle.api.tasks.CacheableTask
-import org.gradle.api.tasks.InputFile
 import org.gradle.api.tasks.OutputDirectory
 import org.gradle.api.tasks.TaskAction
 import java.io.File
 
 @CacheableTask open class DependencyGraphGeneratorTask : DefaultTask() {
   lateinit var generator: Generator // TODO does this need to be an input? Quick testing shows no.
-  @InputFile lateinit var inputFile: File
 
   @OutputDirectory lateinit var outputDirectory: File
 

--- a/src/main/kotlin/com/vanniktech/dependency/graph/generator/ProjectDependencyGraphGeneratorTask.kt
+++ b/src/main/kotlin/com/vanniktech/dependency/graph/generator/ProjectDependencyGraphGeneratorTask.kt
@@ -4,14 +4,12 @@ import com.vanniktech.dependency.graph.generator.DependencyGraphGeneratorExtensi
 import guru.nidi.graphviz.engine.Graphviz
 import org.gradle.api.DefaultTask
 import org.gradle.api.tasks.CacheableTask
-import org.gradle.api.tasks.InputFile
 import org.gradle.api.tasks.OutputDirectory
 import org.gradle.api.tasks.TaskAction
 import java.io.File
 
 @CacheableTask open class ProjectDependencyGraphGeneratorTask : DefaultTask() {
   lateinit var projectGenerator: ProjectGenerator // TODO does this need to be an input? Quick testing shows no.
-  @InputFile lateinit var inputFile: File
 
   @OutputDirectory lateinit var outputDirectory: File
 

--- a/src/test/java/com/vanniktech/dependency/graph/generator/DependencyGraphGeneratorPluginTest.kt
+++ b/src/test/java/com/vanniktech/dependency/graph/generator/DependencyGraphGeneratorPluginTest.kt
@@ -35,7 +35,6 @@ class DependencyGraphGeneratorPluginTest {
     assertThat(task.generator).isSameAs(DependencyGraphGeneratorExtension.Generator.ALL)
     assertThat(task.group).isEqualTo("reporting")
     assertThat(task.description).isEqualTo("Generates a dependency graph")
-    assertThat(task.inputFile).hasToString(singleProject.buildFile.toString())
     assertThat(task.outputDirectory).hasToString(File(singleProject.buildDir, "reports/dependency-graph/").toString())
   }
 
@@ -48,7 +47,6 @@ class DependencyGraphGeneratorPluginTest {
     assertThat(task.projectGenerator).isSameAs(DependencyGraphGeneratorExtension.ProjectGenerator.ALL)
     assertThat(task.group).isEqualTo("reporting")
     assertThat(task.description).isEqualTo("Generates a project dependency graph")
-    assertThat(task.inputFile).hasToString(singleProject.buildFile.toString())
     assertThat(task.outputDirectory).hasToString(File(singleProject.buildDir, "reports/project-dependency-graph/").toString())
   }
 
@@ -138,7 +136,7 @@ class DependencyGraphGeneratorPluginTest {
         }""".trimIndent())
   }
 
-  @Test fun multiProjectIntegrationTest() {
+  @Test @Suppress("Detekt.LongMethod") fun multiProjectIntegrationTest() {
     testProjectDir.newFile("build.gradle").writeText("""
         |plugins {
         |  id "com.vanniktech.dependency.graph.generator"
@@ -150,6 +148,7 @@ class DependencyGraphGeneratorPluginTest {
         |include ":lib1"
         |include ":lib2"
         |include ":app"
+        |include ":empty"
         |""".trimMargin())
 
     val lib = testProjectDir.newFolder("lib").run { parentFile.name + name }
@@ -198,6 +197,8 @@ class DependencyGraphGeneratorPluginTest {
         |}
         |""".trimMargin())
 
+    val empty = testProjectDir.newFolder("empty").run { parentFile.name + name }
+
     val stdErrorWriter = StringWriter()
 
     GradleRunner.create()
@@ -226,9 +227,11 @@ class DependencyGraphGeneratorPluginTest {
         "orgjetbrainskotlinkotlinstdlib" ["shape"="rectangle","label"="kotlin-stdlib"]
         "orgjetbrainsannotations" ["shape"="rectangle","label"="jetbrains-annotations"]
         "$lib2" ["shape"="rectangle","label"="lib2"]
+        "$empty" ["shape"="rectangle","label"="empty"]
         {
         graph ["rank"="same"]
         "$app"
+        "$empty"
         }
         "$app" -> "$lib1"
         "$app" -> "$lib2"

